File: Data Structures & Algorithms/climbing-stairs/submission-2.py
class Solution:
    def climbStairs(self, n: int) -> int:

        memo = defaultdict(int)
        def dfs(i):

            if i == 0:
                return 1
            if i == 1:
                return 1
            if memo[i]:
                return memo[i]

            memo[i] = dfs(i - 1) + dfs(i - 2)
            return memo[i]
     
        # 1, 0
        return dfs(n)

File: Data Structures & Algorithms/count-number-of-islands/submission-2.java
class Solution {
    private int rowSize;
    private int colSize;

    public int numIslands(char[][] grid) {
        rowSize = grid.length;
        colSize = grid[0].length;
        int numOfIslands = 0;

        System.out.println(rowSize);
        System.out.println(colSize);
        for(int r = 0; r < rowSize; r++) {
            for(int c = 0; c < colSize; c++) {
                if (grid[r][c] == '1') {
                    dfs(r, c, grid);
                    numOfIslands++;
                }
            }
        }
        return numOfIslands;
    }

    private void dfs(int r, int c, char[][] grid) {
        
        if (r < 0 || r >= rowSize || c < 0 || c >= colSize || grid[r][c] != '1') return;

        grid[r][c] = '0';
        dfs(r + 1, c, grid);
        dfs(r - 1, c, grid);
        dfs(r, c + 1, grid);
        dfs(r, c - 1, grid);
    }
}
